core/src/main/java/org/apache/gravitino/authorization/PermissionManager.java

```diff
@@ -373,11 +376,174 @@ User revokeRolesFromUser(String metalake, List<String> roles, String user) {
     }
   }
 
-  private List<String> toRoleNames(List<RoleEntity> roleEntities) {
-    return roleEntities.stream().map(RoleEntity::name).collect(Collectors.toList());
+  Role grantPrivilegeToRole(
+      String metalake, String role, MetadataObject object, List<Privilege> privileges) {
+    try {
+      return store.update(
+          AuthorizationUtils.ofRole(metalake, role),
+          RoleEntity.class,
+          Entity.EntityType.ROLE,
+          roleEntity -> {
+            List<SecurableObject> securableObjects = roleEntity.securableObjects();
+
+            boolean objectExist = false;
+
+            List<SecurableObject> updateSecurableObjects = Lists.newArrayList();
+            // Only update the first matched securable object privilege
+            for (SecurableObject securableObject : securableObjects) {
+              if (!objectExist
+                  && securableObject.type() == object.type()
+                  && securableObject.fullName().equals(object.fullName())) {
+                objectExist = true;
+                Set<Privilege> updatePrivileges = Sets.newHashSet();
+                updatePrivileges.addAll(securableObject.privileges());
+                // If privileges has granted, we won't generate the new securable object.
+                if (updatePrivileges.containsAll(privileges)) {
+                  updateSecurableObjects.add(securableObject);
+                } else {
+                  updatePrivileges.addAll(privileges);
+                  SecurableObject newSecurableObject =
+                      SecurableObjects.parse(
+                          securableObject.fullName(),
+                          securableObject.type(),
+                          Lists.newArrayList(updatePrivileges));
+
+                  updateSecurableObjects.add(newSecurableObject);
+                  AuthorizationUtils.callAuthorizationPluginForMetadataObject(
```

Comment:
Should you do this in the SQL transaction, or after the update is done?

Reply:
Done.